File: testdata/fixtures/bench/go/vulnerable/a03_sqli.go
// Source: CWE-89 - SQL Injection in Go web handler
// Expected: GTSS-INJ-001 (SQL Injection via fmt.Sprintf)
// OWASP: A03:2021 - Injection (SQL Injection)

package handler

import (
	"database/sql"
	"fmt"
	"net/http"
)

var db *sql.DB

func SearchUsers(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	sqlQuery := fmt.Sprintf("SELECT id, username, email FROM users WHERE username LIKE '%%%s%%'", query)
	rows, err := db.Query(sqlQuery)
	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var id int
		var username, email string
		rows.Scan(&id, &username, &email)
		fmt.Fprintf(w, "User: %s (%s)\n", username, email)
	}
}

func GetUser(w http.ResponseWriter, r *http.Request) {
	userID := r.URL.Query().Get("id")
	row := db.QueryRow("SELECT username, email FROM users WHERE id = " + userID)
	var username, email string
	if err := row.Scan(&username, &email); err != nil {
		http.Error(w, "Not found", http.StatusNotFound)
		return
	}
	fmt.Fprintf(w, `{"username":"%s","email":"%s"}`, username, email)
}

File: testdata/fixtures/bench/ruby/vulnerable/a03_sqli.rb
# Source: OWASP Rails Goat - SQL injection in ActiveRecord
# Expected: GTSS-INJ-001 (SQL Injection via string interpolation)
# OWASP: A03:2021 - Injection (SQL Injection)

class UsersController < ApplicationController
  def search
    query = params[:q]
    @users = User.where("username LIKE '%#{query}%' OR email LIKE '%#{query}%'")
    render json: @users
  end

  def show
    user_id = params[:id]
    @user = User.find_by_sql("SELECT * FROM users WHERE id = #{user_id}").first
    if @user
      render json: @user
    else
      render json: { error: 'Not found' }, status: 404
    end
  end

  def authenticate
    username = params[:username]
    password = params[:password]
    sql = "SELECT * FROM users WHERE username = '#{username}' AND password_digest = '#{password}'"
    user = ActiveRecord::Base.connection.execute(sql).first
    if user
      session[:user_id] = user['id']
      redirect_to dashboard_path
    else
      flash[:error] = 'Invalid credentials'
      redirect_to login_path
    end
  end
end
